internal/sql/pg/tx.go

```diff
@@ -114,3 +114,65 @@ func (tx *readTx) Rollback(ctx context.Context) error {
 
 	return tx.nestedTx.Rollback(ctx)
 }
+
+// delayedReadTx is a tx that handles a read-only transaction.
+// It is delayed, meaning that the tx will only be actually started
+// when the first query is executed. This is useful for when a calling
+// module is expected to control the lifetime of a read transaction, but
+// the implementation might not need to use the transaction.
+type delayedReadTx struct {
+	db *DB
+
+	tx *readTx
+}
+
+func (d *delayedReadTx) ensureTx() error {
+	if d.tx == nil {
+		tx, err := d.db.BeginReadTx(context.Background())
+		if err != nil {
+			return err
+		}
+
+		d.tx = tx.(*readTx)
+	}
+
+	return nil
+}
+
+func (d *delayedReadTx) Execute(ctx context.Context, stmt string, args ...any) (*common.ResultSet, error) {
+	if err := d.ensureTx(); err != nil {
+		return nil, err
+	}
+
+	return d.tx.Execute(ctx, stmt, args...)
+}
```

Comment:
I believe this transaction becomes abandoned on this path if `ensureTx` had to set `d.tx`.  For example, from `(*TxApp) AccountInfo` where it does `Execute` directly.  It will become an idle transaction in the postgres backend if not committed or rolled back.
Related, that `(*TxApp) AccountInfo` method should be adjusted to take either a `db.Executor` or a `db.TxMaker`.

Reply:
> I believe this transaction becomes abandoned on this path if ensureTx had to set d.tx

I don't think so. The understanding of `delayedReadTx` is that it works like any other tx, and the caller always needs to call Rollback or Commit. Calling Rollback or Commit will always clean this up. The `ensureTx` function is used to effectively "delay" the creation of the original tx, but from the perspective of the consumer, that tx is already open.

Reply:
> Related, that (*TxApp) AccountInfo method should be adjusted to take either a db.Executor or a db.TxMaker. Probably worth checking all the interfaces on the params that take some dbish type.

I'm actually not too sure about this. There aren't many times (if ever) where a `sql.Executor` is not a `sql.TxMaker`, and vice versa. With something as high-level as TxApp, I could very possibly imagine a point in the future where we want transactionality and thus would have to use `sql.DB` (either via type assertion or changing the `AccountInfo` method). 

Overall its pretty minor, and I'm happy to change it if you feel strongly.

Reply:
Have addressed all other changes, let me know if you still want me to make this one.

Reply:
Is the following correct though?

```go
// sendTx makes a transaction and sends it to the local node.
func (svc *Service) sendTx(ctx context.Context, payload transactions.Payload) (*userjson.BroadcastResponse, *jsonrpc.Error) {
	readTx := svc.db.BeginDelayedReadTx()
	defer readTx.Rollback(ctx)

	// Get the latest nonce for the account, if it exists.
	_, nonce, err := svc.TxApp.AccountInfo(ctx, svc.db.BeginDelayedReadTx(), svc.signer.Identity(), true)
```

Did you mean to pass `readTx` to `svc.TxApp.AccountInfo`?

Reply:
> With something as high-level as TxApp, I could very possibly imagine a point in the future where we want transactionality and thus would have to use `sql.DB` (either via type assertion or changing the `AccountInfo` method).

Fair point for the exported methods of `TxApp`, and other high level functions.

Reply:
> Did you mean to pass readTx to svc.TxApp.AccountInfo?

Ah this is a bug. I meant to pass readTx

Reply:
fixed